### ISSUE TITLE
AbstractQuery::toIterable is not a full replacement for AbstractQuery::iterate

### ISSUE THREAD

**Issue by stof** [doctrine/doctrine2#9219]
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | yes (kind of)
| Version     | 2.10.2

#### Summary

The new `AbstractQuery::toIterable` method does not support hydrating mixed results, while the deprecated `AbstractQuery::iterate` was supporting it. This means that it is not always possible to migrate from the deprecated API to its replacement API (and so that some cases would not be possible in ORM 3.0).

It would be great to support it.

**Comment by derrabus** [comment]
Can you provide an example of a piece of code that cannot be migrated to the new API?

**Comment by stof** [comment]
@derrabus any query selecting both an entity result and additional scalar results in its ResultSetMapping (be it in DQL or through `createNativeQuery()`

**Comment by stof** [comment]
This is the code that is explicitly treating it as unsupported: https://github.com/doctrine/orm/blob/7b242753466508e1dd10f67c1baee95785f845c1/lib/Doctrine/ORM/AbstractQuery.php#L1091-L1093

However, removing this condition is not enough, as `\Doctrine\ORM\Internal\Hydration\AbstractHydrator::toIterable` does not support it AFAICT.

**Comment by beberlei** [comment]
Yes this is a known problem, we need to adress this in the generator, but I  haven't found a way yet that does not lead to bugs in edge cases.

**Comment by stof** [comment]
For reference, here is the code I'm using since years in my project (but I don't guarantee that it covers all edge cases):

```php
<?php

namespace Incenteev\WebBundle\Doctrine\Util;

use Doctrine\ORM\AbstractQuery;

class IterationCleaner
{
    /**
     * @phpstan-param AbstractQuery::HYDRATE_* $hydrationMode
     */
    public static function iterateCleanly(AbstractQuery $query, int $hydrationMode): \Traversable
    {
        // Doctrine's Query::iterate API is quite weird: it returns an array of data.
        // And in some cases, it might even split the data into multiple items of the returned array (see https://github.com/doctrine/orm/issues/2821)
        // TODO remove this logic once AbstractQuery::toIterable supports mixed hydration. See https://github.com/doctrine/orm/issues/9219
        foreach ($query->iterate(null, $hydrationMode) as $row) {
            if (count($row) > 1 && is_array($row[0])) {
                yield array_merge(...$row);
            } else {
                yield $row[0];
            }
        }
    }
}
```

**Comment by kbond** [comment]
Related to https://github.com/doctrine/orm/issues/8520. Possible solution: https://github.com/doctrine/orm/pull/8555?

As a workaround, I use the paginator to _chunk_ the results and iterate from each _chunk_.

**Comment by stof** [comment]
This has been fixed in https://github.com/doctrine/orm/pull/12187
